### PR TITLE
Typescript: Convert story editor utils

### LINKED ISSUE #12740: TypeScript: Convert utils folder
## Feature Description

Most of the utility functions inside the `story-editor `package are isolated and only dependent on other already-converted packages. However, many providers inside the `story-editor` package depend on utilities and can't easily be converted unless their utilities are done so as well.

There are a few exceptions to the above though:

* `contrastUtils` depend on the external library [`@ap.cx/hues`](https://www.npmjs.com/package/@ap.cx/hues) which does not appear to be typed or have types available
* `getUpdatedSizeAndPosition` and `presetUtils` depend on un-converted functions inside the `element-library` package
* `storyPageToCanvas` and `storyPageToDataUrl` depend on `storyPageToNode`, and
* `storyPageToNode` depends on the `font` provider and the canvas component
* `useAddPreset` depends on the `story` provider
* `useApplyTextAutoStyle` depends on the `pageCanvas` provider and the style panel component
* `useCORSProxy` depends on the `api` provider
* `useElementPolygon` depends on the `story` and `canvas` providers
* `useElementWithLink` depends on the `story` and `canvas` providers
* `useIsUploadingToStory` depends on the `story` and `localMedia` providers
* `useShapeMask` depends on the `story` and `canvas` providers
* `useShapeMaskElements` depends on the `story` provider

These functions should be moved to their respective providers (mostly to live inside the story provider) and or other relevant parts of the codebase, or their dependency types should be mocked.

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Covert the simple utils in story editor to use type script. 

## Relevant Technical Choices

The following files were not converted, for these reasons. 
- contrastUtils.js - Can’t convert as not typed @ap.cx/hues'
- copyPaste.js - Too complex and lots of lints errors that could not be resolved. 
- generateBlurhash.worker.js / getBlurHashFromImage.js - Can’t convert because used in worker. Requires changes to build process to get working. 
- getUpdatedSizeAndPosition.js - Can’t convert as blocked by calculateTextHeight
- useAddPreset.js - useStory not converted. 
- useApplyTextAutoStyle.js - blocked by useCalculateAccessibleTextColors
- useElementPolygon.js - blocked by canvas
- useElementsWithLinks - blocked by canvas
- useFocusTrapping - useKeyDownEffect params or incorrect
- useIsUploadingToStory - blocked by upload
- useShapeMask - useStory not converted. 
- useShapeMaskElements - useStory not converted. 

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12740

### CODE REVIEW

File: packages/story-editor/src/utils/useGlobalClipboardHandlers.ts [hunk 1]
```diff
@@ -25,21 +25,15 @@ import { useEffect } from '@googleforcreators/react';
...
- * @param {ClipboardEvent} event
- */
-
-/**
- * @param {CopyCutHandler} copyCutHandler Copy & Cut handler.
- * @param {PasteHandler} pasteHandler Paste handler.
- */
-function useGlobalClipboardHandlers(copyCutHandler, pasteHandler) {
+function useGlobalClipboardHandlers(
+  copyCutHandler: (event: ClipboardEvent) => void,
```
Comment (miina) [inline packages/story-editor/src/utils/useGlobalClipboardHandlers.ts]: Nit: Could create a separate `type` for it since there's two of the same.

File: packages/story-editor/src/utils/storyUpdates.ts [hunk 2]
```diff
@@ -22,20 +22,35 @@
...
  */
-export const updateSlug = ({ currentSlug, currentTitle, updateStory }) => {
+export const updateSlug = ({
+  currentSlug,
+  currentTitle,
+  updateStory,
+}: updateSlugProps) => {
   if (!currentSlug) {
-    const cleanSlug = encodeURIComponent(cleanForSlug(currentTitle));
+    const cleanSlug = encodeURIComponent(String(cleanForSlug(currentTitle)));
```
Comment (miina) [inline packages/story-editor/src/utils/storyUpdates.ts]: Shouldn't `cleanForSlug` have `string` as a return type already?

File: packages/story-editor/src/utils/storyUpdates.ts [hunk 3]
```diff
@@ -22,20 +22,35 @@
...
+interface updateStoryProps {
+  properties: {
+    slug: string;
+  };
+}
+
+interface updateSlugProps {
+  currentSlug: string;
+  currentTitle: string;
+  updateStory: (props: updateStoryProps) => void;
```
Comment (miina) [inline packages/story-editor/src/utils/storyUpdates.ts]: See the `updateStory` in the `StoryProvider`, these types should probably match since the `updateStory` comes from the story provider.

File: packages/story-editor/src/utils/nativeCopyPasteExpected.ts [hunk 4]
```diff
@@ -14,28 +14,38 @@
...
-  if (
-    'true' === contentEditable ||
-    'textarea' === tagName.toLowerCase() ||
-    ('input' === tagName.toLowerCase() &&
-      ['text', 'number', 'search', 'email', 'tel', 'url'].includes(type))
-  ) {
-    return true;
-  }
+  if (activeElement) {
+    const { tagName, type, contentEditable } = activeElement as ElementProps;
```
Comment (miina) [inline packages/story-editor/src/utils/nativeCopyPasteExpected.ts]: There are default types for elements that have these values, let's use those.
E.g. the activeElement could be a union type, something like
`HTMLInputElement
  | HTMLAnchorElement
  | HTMLTextAreaElement` perhaps?

File: packages/story-editor/src/utils/isDefaultPage.ts [hunk 5]
```diff
@@ -40,18 +37,30 @@ const isDefaultPage = (page) => {
...
 
   // Check if background element is not default
-  const backgroundElement = page.elements.find((e) => e.isBackground);
-  if (backgroundElement && !backgroundElement.isDefaultBackground) {
-    return false;
+  const backgroundElement = page.elements.find(
+    (e) => elementIs.backgroundable(e) && e.isBackground
+  );
+  if (backgroundElement) {
+    if (
```
Comment (miina) [inline packages/story-editor/src/utils/isDefaultPage.ts:52]: Maybe combine these two `if`s.

File: packages/story-editor/src/utils/isDefaultPage.ts [hunk 6]
```diff
@@ -40,18 +37,30 @@ const isDefaultPage = (page) => {
...
   if (
-    !shallowEqual(
-      page.backgroundColor?.color,
-      defaultPage.backgroundColor?.color
-    )
+    'color' in page.backgroundColor &&
+    'color' in defaultPage.backgroundColor
   ) {
-    return false;
+    if (
```
Comment (miina) [inline packages/story-editor/src/utils/isDefaultPage.ts]: Here, too, lets combine these two `if`s. Alternatively we could return after the first if already. Nested if's like that without any content in between aren't the best. Eg. 
```
if (
!'color' in page.backgroundColor ||
    !'color' in defaultPage.backgroundColor
) { 
return false;
}
// Second if now.
```

File: packages/story-editor/src/utils/idleCallback.ts [hunk 7]
```diff
@@ -17,15 +17,15 @@
...
+  function (callback: IdleRequestCallback, options?: IdleRequestOptions) {
     options = options || {};
     const relaxation = 1;
     const timeout = options.timeout || relaxation;
     const start = performance.now();
     return setTimeout(() => {
       callback({
         get didTimeout() {
-          return options.timeout
+          return options?.timeout
```
Comment (miina) [inline packages/story-editor/src/utils/idleCallback.ts]: Since `options = options || {};` then `options` should always exist though, no?

File: packages/story-editor/src/utils/getInUseFonts.ts [hunk 8]
```diff
@@ -40,23 +44,31 @@ export function getInUseFontsForPages(pages) {
...
+  textSets,
+}: getTextSetsForFontsProps) {
   return textSets
     .map((currentTextSet) => {
       const hasFontInUse = currentTextSet.elements.reduce(
         (elementMemo, currentElement) => {
           if (
-            currentElement.type === 'text' &&
+            elementIs.text(currentElement) &&
             Boolean(currentElement.font?.family)
```
Comment (miina) [inline packages/story-editor/src/utils/getInUseFonts.ts]: Since `font.family` is required in a text element then this additional check shouldn't be necessary.

File: packages/story-editor/src/utils/getInUseFonts.ts [hunk 9]
```diff
@@ -14,20 +14,24 @@
...
-export function getInUseFontsForPages(pages) {
+export function getInUseFontsForPages(pages: Page[]) {
   return (
     Array.from(
       new Set(
         pages
           .map(({ elements = [] }) =>
             elements.map((element) => {
-              if (element.type === 'text' && Boolean(element.font?.family)) {
+              if (elementIs.text(element) && Boolean(element.font?.family)) {
```
Comment (miina) [inline packages/story-editor/src/utils/getInUseFonts.ts]: Same here, `font.family` is required for text element.

File: packages/story-editor/src/utils/generateGroupName.ts [hunk 10]
```diff
@@ -19,28 +19,39 @@
...
+    const parts = name.split(' ');
     if (parts[0] === defaultName && parts[1]) {
       nums.push(Number(parts[1]));
     }
   }
   return Math.max(...nums) + 1;
 }
 
-function generateGroupName(groups, name) {
+function generateGroupName(groups: GroupsProps, name: string): string {
```
Comment (miina) [inline packages/story-editor/src/utils/generateGroupName.ts]: Hm, is the return type not inferred?

File: packages/story-editor/src/utils/generateGroupName.ts [hunk 11]
```diff
@@ -19,28 +19,39 @@
...
   const defaultName = __('Group', 'web-stories');
   for (const prop in groups) {
     if (!Object.prototype.hasOwnProperty.call(groups[prop], 'name')) {
       continue;
     }
-    const parts = groups[prop]?.name.split(' ');
+    const { name = '' } = groups[prop];
+    if (!name) {
+      continue;
+    }
```
Comment (miina) [inline packages/story-editor/src/utils/generateGroupName.ts]: Maybe just
`if (!groups[prop]?.name) return;`?

File: packages/story-editor/src/utils/generateGroupName.ts [hunk 12]
```diff
@@ -19,28 +19,39 @@
...
 import { __, sprintf } from '@googleforcreators/i18n';
 
-export function getNextGroupNumber(groups) {
+interface GroupsProps {
+  [key: string]: {
+    name?: string;
+  };
+}
+
+export function getNextGroupNumber(groups: GroupsProps) {
```
Comment (miina) [inline packages/story-editor/src/utils/generateGroupName.ts]: You could just import `Groups` from the `elements` package.

File: packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.js [hunk 13]
```diff
@@ -52,7 +52,10 @@ function PageDataUrlProvider({ children }) {
         }));
       };
 
-      const clearQueueOfPageTask = queueIdleTask([idleTaskUid, idleTask]);
+      const clearQueueOfPageTask = queueIdleTask({
```
Comment (miina) [inline packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.js:55]: Curious why the change?

File: packages/elements/src/types/elementDefinition.ts [hunk 14]
```diff
@@ -90,6 +90,12 @@ export interface ElementDefinition<E extends Element = Element> {
     minHeight: number;
   };
   panels: string[];
+  updateForResizeEvent?: (
+    element: Element,
+    direction: number[],
```
Comment (miina) [inline packages/elements/src/types/elementDefinition.ts]: A tuple would be more accurate here.

File: packages/story-editor/src/utils/generateGroupName.ts (hunk 11 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/generateGroupName.ts]: I can try it, but I tried a number of these and got a lint error everytime.

File: packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.js (hunk 13 above)
Comment (spacedmonkey) [reply packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.js:55]: It makes the code easier to understand and type.

File: packages/story-editor/src/utils/storyUpdates.ts (hunk 2 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/storyUpdates.ts]: Screenshot 2022-12-07 at 10 23 16

**Comment by swissspidy** [comment]
Looks like you need to rebase this again, there's now a lot of changes from `main` in this PR

File: packages/story-editor/src/utils/generateGroupName.ts (hunk 12 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/generateGroupName.ts]: Fixed

File: packages/story-editor/src/utils/generateGroupName.ts (hunk 11 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/generateGroupName.ts]: Removed

File: packages/story-editor/src/utils/getInUseFonts.ts (hunk 9 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/getInUseFonts.ts]: Fixed.

File: packages/story-editor/src/utils/getInUseFonts.ts (hunk 8 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/getInUseFonts.ts]: Fixed.

File: packages/story-editor/src/utils/idleCallback.ts (hunk 7 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/idleCallback.ts]: But the property of timeout may not

File: packages/story-editor/src/utils/generateGroupName.ts [hunk 15]
```diff
@@ -18,29 +18,31 @@
...
+export function getNextGroupNumber(groups: Groups) {
   const nums = [0];
   const defaultName = __('Group', 'web-stories');
   for (const prop in groups) {
     if (!Object.prototype.hasOwnProperty.call(groups[prop], 'name')) {
       continue;
     }
-    const parts = groups[prop]?.name.split(' ');
+
+    const parts = groups[prop].name.split(' ');
```
Comment (spacedmonkey) [inline packages/story-editor/src/utils/generateGroupName.ts]:
```suggestion
    const parts = groups[prop]?.name.split(' ');
```

File: packages/story-editor/src/utils/idleCallback.ts (hunk 7 above)
Comment (miina) [reply packages/story-editor/src/utils/idleCallback.ts]: Yes, but the change doesn't influence that -- the added `?` has no effect related to the `timeout` prop being undefined or not. And since `options` always exist, it doesn't do anything basically.

File: packages/story-editor/src/utils/storyUpdates.ts (hunk 2 above)
Comment (miina) [reply packages/story-editor/src/utils/storyUpdates.ts]: Regardless, `cleanForSlug` should have `string` as a return type, no? Maybe `cleanForSlug` needs an update then, too.

File: packages/story-editor/src/utils/noop.ts [hunk 16]
```diff
@@ -13,4 +13,5 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+// eslint-disable-next-line @typescript-eslint/no-empty-function -- Known.
 export function noop() {}
```
Comment (swissspidy) [inline packages/story-editor/src/utils/noop.ts]:
```suggestion
export function noop() {
  return undefined;
}
```

File: packages/story-editor/src/utils/inRange.ts [hunk 17]
```diff
@@ -13,6 +13,13 @@
...
  * limitations under the License.
  */
-const inRange = (value, { MIN, MAX }) => value >= MIN && value <= MAX;
+
+interface minMaxProps {
+  MIN: number;
+  MAX: number;
+}
+
+const inRange = (value: number, { MIN, MAX }: minMaxProps) =>
```
Comment (swissspidy) [inline packages/story-editor/src/utils/inRange.ts]: While we're at it:

This function is only ever used in the `SlugPanel` component and it's basically a single line of code.

Let's just delete this file and define that function inline in that component's file.

File: packages/elements/src/types/elementDefinition.ts (hunk 14 above)
Comment (spacedmonkey) [reply packages/elements/src/types/elementDefinition.ts]: Fixed.

Comment (spacedmonkey) [reply packages/elements/src/types/elementDefinition.ts]: Fixed.

File: packages/story-editor/src/utils/storyUpdates.ts (hunk 2 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/storyUpdates.ts]: Fixed in https://github.com/GoogleForCreators/web-stories-wp/pull/12737/commits/0648200ccafe4d71057a9c37a2355e9ba7e3ebb1

Comment (spacedmonkey) [reply packages/story-editor/src/utils/storyUpdates.ts]: Fixed in https://github.com/GoogleForCreators/web-stories-wp/pull/12737/commits/0648200ccafe4d71057a9c37a2355e9ba7e3ebb1

File: packages/story-editor/src/utils/nativeCopyPasteExpected.ts (hunk 4 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/nativeCopyPasteExpected.ts]: Fixed.

File: packages/story-editor/src/utils/getCropParams.ts [hunk 18]
```diff
@@ -13,12 +13,16 @@
...
+ * External dependencies
+ */
+import type { MediaElement } from '@googleforcreators/elements';
 /**
  * Internal dependencies
  */
 import { isOffCanvas } from './isOffCanvas';
 
-export function getCropParams(selectedElement) {
+export function getCropParams(selectedElement: MediaElement) {
```
Comment (swissspidy) [inline packages/story-editor/src/utils/getCropParams.ts]: Just noting that in #12763 I defined a new `CropParams` type that could be used here for the return value.

File: packages/story-editor/src/app/pageDataUrls/test/usePageDataUrls.js [hunk 19]
```diff
@@ -60,19 +60,19 @@ describe('usePageDataUrls', () => {
     mockIdleCallbacks = [];
     runIdleCallbacks = () => {
       while (mockIdleCallbacks.length > 0) {
-        const [, callback] = mockIdleCallbacks.shift();
-        callback();
+        const { task } = mockIdleCallbacks.shift();
```
Comment (swissspidy) [inline packages/story-editor/src/app/pageDataUrls/test/usePageDataUrls.js:63]: Just noting that converting this test to TypeScript could help with these changes here, making the test more robust

File: packages/story-editor/src/utils/storyUpdates.ts [hunk 20]
```diff
@@ -20,20 +20,30 @@
...
  *
- * @param {Object} args Necessary data for update.
- * @param {string} args.currentSlug The currently assigned value to story's slug.
- * @param {string} args.currentTitle The currently assigned value to story's title.
- * @param {Function} args.updateStory The callback to useStory's action for updating the story with new slug.
- * @return {void}
+ * @param args Necessary data for update.
+ * @param args.currentSlug The currently assigned value to story's slug.
+ * @param args.currentTitle The currently assigned value to story's title.
+ * @param args.updateStory The callback to useStory's action for updating the story with new slug.ß
```
Comment (swissspidy) [inline packages/story-editor/src/utils/storyUpdates.ts]:
```suggestion
 * @param args.updateStory The callback to useStory's action for updating the story with new slug.
```

File: packages/story-editor/src/utils/pointerEventsCss.ts [hunk 21]
```diff
@@ -19,8 +19,12 @@
  */
 import { css } from 'styled-components';
 
+interface pointerEventsCssProps {
+  pointerEvents?: string;
```
Comment (swissspidy) [inline packages/story-editor/src/utils/pointerEventsCss.ts]:
```suggestion
  pointerEvents?: 'none' | 'auto' | 'initial';
```

Also makes the `typeof pointerEvents === 'string'` check below superfluous.

File: packages/story-editor/src/utils/useCORSProxy.ts [hunk 22]
```diff
@@ -64,9 +64,9 @@ function useCORSProxy() {
    * @return {null|string} Return proxied source or null.
    */
   const getProxiedUrl = useCallback(
-    (resource, src) => {
+    (resource: Resource, src: string) => {
       const { needsProxy } = resource;
-      if (needsProxy && src) {
+      if (needsProxy && src && typeof getProxyUrl === 'function') {
```
Comment (swissspidy) [inline packages/story-editor/src/utils/useCORSProxy.ts]:
```suggestion
      if (needsProxy && src && getProxyUrl) {
```

This should be enough.

See similar change in #12763

File: packages/story-editor/src/utils/useDoubleClick.ts [hunk 23]
```diff
@@ -19,6 +19,8 @@
  */
 import { useCallback, useEffect, useState } from '@googleforcreators/react';
 
+type clickCallback = (evt: Event, target: Element) => void;
```
Comment (swissspidy) [inline packages/story-editor/src/utils/useDoubleClick.ts]: Nit: let's capitalize our types for consistency. I'll see if we can enforce this with the linter

Also, can we perhaps narrow down the type a bit more?

This should be a `MouseEvent`.

Perhaps can use `MouseEventHandler<HTMLElement>` too

File: packages/story-editor/src/utils/idleCallback.ts (hunk 7 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/idleCallback.ts]:
```suggestion
          return options.timeout
```

File: packages/story-editor/src/app/pageDataUrls/test/usePageDataUrls.js (hunk 19 above)
Comment (spacedmonkey) [reply packages/story-editor/src/app/pageDataUrls/test/usePageDataUrls.js:63]: Should we convert the tests in these PRs?

File: packages/story-editor/src/utils/idleCallback.ts (hunk 7 above)
Comment (swissspidy) [reply packages/story-editor/src/utils/idleCallback.ts]: Reopening this because TS actually thinks `options` can be `undefined`.

To resolve, I propose using `function (callback: IdleRequestCallback, options: IdleRequestOptions = {}) {`

File: packages/story-editor/src/utils/useDoubleClick.ts (hunk 23 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/useDoubleClick.ts]: Fixed.

File: packages/story-editor/src/utils/getCropParams.ts [hunk 24]
```diff
@@ -13,12 +13,30 @@
...
+ * External dependencies
+ */
+import type { MediaElement } from '@googleforcreators/elements';
 /**
  * Internal dependencies
  */
 import { isOffCanvas } from './isOffCanvas';
 
-export function getCropParams(selectedElement) {
+// TODO, reuse CropParams from media utils.
```
Comment (swissspidy) [inline packages/story-editor/src/utils/getCropParams.ts]: This is now merged and available 👍 

Maybe need the type definition to a better location.

File: packages/story-editor/src/utils/useIdleTaskQueue.ts [hunk 25]
```diff
@@ -23,15 +23,20 @@ import { useRef, useCallback } from '@googleforcreators/react';
  */
 import { requestIdleCallback, cancelIdleCallback } from './idleCallback';
 
+interface Task {
+  taskId: string | null;
+  task: number | null | (() => Promise<void>);
```
Comment (barklund) [inline packages/story-editor/src/utils/useIdleTaskQueue.ts:28]: Why can task be a number? What would that be used for?

File: packages/story-editor/src/utils/getInUseFonts.ts [hunk 26]
```diff
@@ -40,26 +41,32 @@ export function getInUseFontsForPages(pages) {
   );
 }
 
+interface getTextSetsForFontsProps {
```
Comment (barklund) [inline packages/story-editor/src/utils/getInUseFonts.ts]:
```suggestion
interface GetTextSetsForFontsProps {
```

File: packages/story-editor/src/utils/getInsertedElementSize.ts [hunk 27]
```diff
@@ -24,28 +24,31 @@ import {
...
+const RESIZE_WIDTH_DIRECTION: Direction = [1, 0];
 
 /**
- * @param {?number|undefined} value The value.
- * @return {boolean} Whether the value has been set.
+ * @param The value.
+ * @return  Whether the value has been set.
  */
-function isNum(value) {
+function isNum(value: unknown) {
```
Comment (barklund) [inline packages/story-editor/src/utils/getInsertedElementSize.ts]: The return type here could even be a type guard:

```suggestion
function isNum(value: unknown): value is number {
```

Not sure if it makes a difference though.

File: packages/story-editor/src/utils/isOffCanvas.ts [hunk 28]
```diff
@@ -21,8 +21,9 @@ import {
   PAGE_HEIGHT,
   PAGE_WIDTH,
 } from '@googleforcreators/units';
+import type { Element } from '@googleforcreators/elements';
 
-export function isOffCanvas(selectedElement) {
+export function isOffCanvas(selectedElement: Element) {
```
Comment (barklund) [inline packages/story-editor/src/utils/isOffCanvas.ts]: If only x, y, width and height are used, this can be typed more loosely as `ElementBox` from the unit package.

File: packages/story-editor/src/utils/nativeCopyPasteExpected.ts [hunk 29]
```diff
@@ -18,24 +18,31 @@
...
-    'textarea' === tagName.toLowerCase() ||
-    ('input' === tagName.toLowerCase() &&
-      ['text', 'number', 'search', 'email', 'tel', 'url'].includes(type))
-  ) {
-    return true;
-  }
+  if (activeElement) {
+    const { tagName, type, contentEditable } = activeElement as
+      | HTMLInputElement
+      | HTMLAnchorElement
```
Comment (barklund) [inline packages/story-editor/src/utils/nativeCopyPasteExpected.ts:29]: Why anchor elements? `contentEditable` can be applied to any element.

File: packages/story-editor/src/utils/storyUpdates.ts [hunk 30]
```diff
@@ -20,20 +20,30 @@
 /**
  * Internal dependencies
  */
+import type { ReducerState, UpdateStoryProps } from '../types';
 import cleanForSlug from './cleanForSlug';
 
+interface updateSlugProps {
```
Comment (barklund) [inline packages/story-editor/src/utils/storyUpdates.ts]:
```suggestion
interface UpdateSlugProps {
```

File: packages/story-editor/src/utils/useDoubleClick.ts [hunk 31]
```diff
@@ -50,16 +52,20 @@ import { useCallback, useEffect, useState } from '@googleforcreators/react';
  * }
  * ```
  *
- * @param {Function} onSingleClick  Handler to activate on single click.
- * @param {Function} onDoubleClick  Handler to activate on double click.
- * @param {number}   ms             Timeout in ms to wait - defaults to 200.
- * @return {Function} Handler retrieval function to get an onClick listener (invoke with unique value).
+ * @param onSingleClick  Handler to activate on single click.
+ * @param onDoubleClick  Handler to activate on double click.
+ * @param   ms             Timeout in ms to wait - defaults to 200.
```
Comment (barklund) [inline packages/story-editor/src/utils/useDoubleClick.ts]:
```suggestion
 * @param ms             Timeout in ms to wait - defaults to 200.
```

**Review by barklund — APPROVED** [review]
Looks good and sorely needed in many tickets. 🧨

File: packages/story-editor/src/utils/useIdleTaskQueue.ts (hunk 25 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/useIdleTaskQueue.ts:28]: Because it changes from a promise to `idleCallbackId`, which is an int. That int is then passed to `clearQueuedTask` which expects an int. 

Really hard to follow, I know, took me a while to understand.

File: packages/story-editor/src/utils/nativeCopyPasteExpected.ts (hunk 29 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/nativeCopyPasteExpected.ts:29]: Recommended by @miina

Comment (miina) [reply packages/story-editor/src/utils/nativeCopyPasteExpected.ts:29]: @spacedmonkey The idea was that previously you had created a custom interface to support `tagName, type, contentEditable`, however, [I mentioned there are existing default types](https://github.com/GoogleForCreators/web-stories-wp/pull/12737#discussion_r1039984914) already. I did not look into it thoroughly though, so there might be some missing.

@barklund I don't think that `contentEditable` and `type` both exists on `Element` / `HTMLElement` so this needs more precision. In this particular case, we are trying to detect if native copy-paste should take over or not.

Comment (spacedmonkey) [reply packages/story-editor/src/utils/nativeCopyPasteExpected.ts:29]: The typing is hard here. It can be multiple types, that is valid. Not all types have the same props, some have contentEditable and other have type. Not sure what to do here other than what as been done. Maybe a custom type.

File: packages/story-editor/src/utils/useDoubleClick.ts (hunk 31 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/useDoubleClick.ts]: Fixed

File: packages/story-editor/src/utils/storyUpdates.ts (hunk 30 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/storyUpdates.ts]: Fixed.

File: packages/story-editor/src/utils/isOffCanvas.ts (hunk 28 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/isOffCanvas.ts]: Fixed.

File: packages/story-editor/src/utils/getInsertedElementSize.ts (hunk 27 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/getInsertedElementSize.ts]: Not sure what value this would add here...

File: packages/story-editor/src/utils/getInUseFonts.ts (hunk 26 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/getInUseFonts.ts]: Fixed.

File: packages/story-editor/src/utils/getCropParams.ts (hunk 24 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/getCropParams.ts]: Fixed.

File: packages/story-editor/src/utils/getInsertedElementSize.ts (hunk 27 above)
Comment (swissspidy) [reply packages/story-editor/src/utils/getInsertedElementSize.ts]: [Type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) like this are useful for type narrowing

File: packages/story-editor/src/utils/getCropParams.ts [hunk 32]
```diff
@@ -13,12 +13,18 @@
...
  */
+/**
+ * External dependencies
+ */
+import type { MediaElement } from '@googleforcreators/elements';
+
 /**
  * Internal dependencies
  */
+import type { CropParams } from '../app/media';
```
Comment (swissspidy) [inline packages/story-editor/src/utils/getCropParams.ts]:
```suggestion
import type { CropParams } from '../app/media/types';
```

Otherwise when importing from `../app/media/` TS will complain about all of these files there not being typed.

File: packages/story-editor/src/utils/isOffCanvas.ts [hunk 33]
```diff
@@ -21,8 +21,9 @@ import {
   PAGE_HEIGHT,
   PAGE_WIDTH,
 } from '@googleforcreators/units';
+import type { ElementBox } from '@googleforcreators/elements';
```
Comment (swissspidy) [inline packages/story-editor/src/utils/isOffCanvas.ts]: `ElementBox` is in the `units` package, not the `elements` package. You should see a warning about this in your IDE and by TS.

**Review by swissspidy — COMMENTED** [review]
Looks good apart from the few open suggestions that need resolving

File: packages/story-editor/src/utils/getCropParams.ts (hunk 32 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/getCropParams.ts]: Originally had it like this. The linter changed it.

File: packages/story-editor/src/utils/getInsertedElementSize.ts [hunk 34]
```diff
@@ -24,28 +24,27 @@ import {
...
-  width,
-  height,
-  attrs,
+  type: ElementType,
+  width: number,
+  height: number,
+  attrs: object,
   ratio = 1,
-  resource = null
+  resource: Resource
```
Comment (swissspidy) [inline packages/story-editor/src/utils/getInsertedElementSize.ts]:
```suggestion
  resource?: Resource
```

Since it was basically optional before and a resource is not always provided (for example `useInsertPreset` does not pass a resource)

Comment (spacedmonkey) [reply packages/story-editor/src/utils/getInsertedElementSize.ts]: Fixed in https://github.com/GoogleForCreators/web-stories-wp/pull/12737/commits/6fd2808d4c2c96bf41c8807940aa91c85753e4dd

**Review by swissspidy — APPROVED** [review]
